### DISCUSSION TITLE
Java Language Server Startup

### DISCUSSION THREAD

**Discussion by bdresdner** [eclipse-theia/theia#12947] [Q&A] [answered]
Hi, I know this question probably belongs in a redhat-developer/vscode-java discussion, but I posted it there and have not received any responses.  I have Theia running in a k8s pod with limited CPU memory.  It takes long time for the Java language server to start after editing a Java class.  What I am attempting to do is start the Java Language server when the pod starts so that it is up and running before the user opens a Java class in the UI.  I used the Python wrapper script as documented [here](https://github.com/eclipse-jdtls/eclipse.jdt.ls#running-from-command-line-with-wrapper-script).  The command looks as follows:

`/home/sag/node_modules/@jse/java-service-editor-app/plugins/redhat.java/extension/server/bin/jdtls -configuration ~/.cache/jdtls -data /home/project`

Where /home/project is the location of my workspace.
/home/sag/node_modules/@jse is my Theia extension.

The command to start the language server appears to work, but when I go into the UI for my Theia extension, I do not see any feedback in the status bar that the language server is running.  When I edit a Java class in my workspace, I then see it start.

Can anyone tell me what I may be doing wrong or if this is even possible?

**Comment by msujew** [comment, marked as answer]
@bdresdner Language Servers are not actual servers, in the sense that they start up on some port and accept requests/return responses. Instead, they are designed so that each workspace starts up it's own language server process with which it communicates via IPC. Which is why the java vscode extension will start up it's own language server instance once it it's `activate` function will be called (which will only happen in case a `.java` file is opened by the user).

What you could do in theory, is to immediately start the java extension once the plugin host starts up. However, the plugin host only starts once a user connects to the backend, so it won't be a dramatic startup time increase. It only skips the time between \<user connects\> and \<user opens a Java file\>. Every other optimization is kind of limited by the general Theia/VSCode architecture.

**Reply by bdresdner** [reply]
Hi @msujew, thank you for your detailed response.  This helps s lot.  I will try your suggestion.

**Reply by msujew** [reply]
A quick way to accomplish this would be to replace the [activationEvents](https://github.com/redhat-developer/vscode-java/blob/812613b5923ba27508f0a15a605f4cf80faa2388/package.json#L42-L56) of the extension just with `"*"`, which leads the extension to start immediately.

**Reply by bdresdner** [reply]
Do you mean the activationEvents of my custom extension?  Currently, I have none.  If I understand correctly, I should add the following to my package.json:
`"activationEvents": *,`
Is that right?

**Reply by msujew** [reply]
Take a look at the linked package.json. A vscode extension is able to define when the IDE is supposed to call it's `activate` function. In the case of the java extension, it's the linked lines. You can replace all of those with just `"*"` or `"onStartup"` to indicate that vscode/Theia is supposed to start the extension as soon as possible.

Of course you could also instruct Theia in your custom Theia extension to just activate the extension. That works as well.

**Reply by bdresdner** [reply]
Thanks, I am still trying to make sure I am clear on this.  It sounds like I would have to modify the package.json of the java extension.  Currently, the package.json of my custom extension downloads the  java extension every time it runs.  Does this mean that I should manually download the java extension and modify its package.json then have my build use the modified extension?

**Reply by msujew** [reply]
I'm not really recommending that. I just said it's a quick solution, not a maintenable one. The clean solution would be to indeed use a Theia extension to invoke the corresponding plugin activation.

> Currently, the package.json of my custom extension downloads the java extension every time it runs.

That sounds pretty wasteful. Depending on your deployment method (electron/browser) you should bake the extension into the app image (electron zip/docker image).

**Reply by bdresdner** [reply]
@msujew Thank you.  Would you happen to know of a snippet from a Theia extension that invokes plugin activation?
BTW - My package.json is to similar [Theia Blueprint](https://github.com/eclipse-theia/theia-blueprint/blob/master/package.json)'s.

**Reply by msujew** [reply]
You'll need to call the `activateByEvent` function somehow with the expected event. See [here](https://github.com/eclipse-theia/theia/blob/ba3722b04ff91eb6a4af6a571c9e263c77cdd8b5/packages/plugin-ext/src/main/browser/plugin-authentication-service.ts#L41) for an example.

> BTW - My package.json is to similar [Theia Blueprint](https://github.com/eclipse-theia/theia-blueprint/blob/master/package.json)'s.

Uhm, ok. Theia Blueprint doesn't re-download its builtin plugins on startup. It just performs the `download:plugins` script once on build and that's it. No need to redownload the builtins on every startup. At least that's how I understood [your comment](https://github.com/eclipse-theia/theia/discussions/12947#discussioncomment-7122685).

**Reply by bdresdner** [reply]
Thank you for all your help!  I learned a lot.  I'll try your suggestion and let you know how it works out.

**Reply by bdresdner** [reply]
I am calling `activateByEvent` passing` "onLanguage:java"`, but it is not starting the Java language server.  Thanks again, it was worth a try.
